teps/0044-data-locality-and-pod-overhead-in-pipelines.md

```diff
@@ -241,6 +243,40 @@ However, the number of Tasks needed may not be known until the previous Task is
 Therefore, we may be able to support running a matrixed Pipeline in a pod only when the full set of parameters is known
 at the start of execution. We may not be able to support dynamic matrix parameters or other forms of dynamic Task creation.
 
+### Hermekton support
+
+[TEP-0025](./0025-hermekton.md) proposes specifying "hermeticity" levels for Tasks or Steps. Hermetic Tasks and Steps cannot
+communicate with non-hermetic Tasks and Steps during execution, meaning that all inputs will be specified prior to Task/Step start.
+This requires isolating the network and filesystem for hermetic Tasks/Steps.
+
+#### Use Cases
+
+TEP-0025 describes some [use cases](./0025-hermekton.md#user-stories-optional) for specifying different levels of hermeticity for
+Tasks in a Pipeline. It's not yet clear whether users would like the ability to specify different levels of hermeticity for Tasks
+that are part of a PipelineRun executed in a pod. Security-minded users might prefer to run their Tasks with different service accounts,
+which is not possible in one pod.
+
+#### Feasibility
+
+A Task can only be considered hermetic if it is unable to communicate with other Tasks during execution, and likewise for Steps.
+If we wanted to support different levels of hermeticity for Tasks run in the same pod, we would need to provide a way for the
+Steps in the hermetic Task to communicate with other Steps in that Task, but not with Steps in other Tasks.
+
+Containers in a pod can [communicate](https://kubernetes.io/docs/concepts/workloads/pods/#resource-sharing-and-communication)
+either via ports or via the filesystem.
+
+Isolating the filesystem of a Task run in the same pod as other Tasks is likely feasible.
+A pod can provide shared volumes for containers to use, meaning that we can control how containers communicate via the filesystem
+by controlling which of the pod's volumes they have access to.
+
+Isolating the network of a Task run in the same pod as other Tasks is more challenging.
+Containers in a pod share a network namespace, and can communicate with each other via localhost.
+This means that it's straightforward to restrict a single container's access to the network within the pod by preventing it from
+communicating via localhost, as is proposed for [Step-level hermeticity](./0025-hermekton.md#api-changes).
+However, it's much more challenging to allow a group of containers to communicate with each other, but not with other containers in a pod,
```

Comment:
what if the hermetic support within pipeline in a pod was limited to Tasks with only one step? would it be feasible to mix hermetic and non-hermetic Tasks if we could guarantee there was only one step in each hermetic Task (i.e. it had no need to communicate with any other steps)?

or possibly it was required that the steps within a Task that is run hermetically inside a pod not communicate with each other? (probably a less good option since why would you have multiple steps at all unless they are communicating with each other - however we recently added [per step hermetic execution](https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md#api-changes) to the hermekton TEP which makes me think there is some value in having multiple steps even when they can't communicate with each other?)

Reply:
> what if the hermetic support within pipeline in a pod was limited to Tasks with only one step? would it be feasible to mix hermetic and non-hermetic Tasks if we could guarantee there was only one step in each hermetic Task (i.e. it had no need to communicate with any other steps)?

Can multiple Tasks run in parallel in "pipeline in a pod" mode?

Reply:
> what if the hermetic support within pipeline in a pod was limited to Tasks with only one step? would it be feasible to mix hermetic and non-hermetic Tasks if we could guarantee there was only one step in each hermetic Task (i.e. it had no need to communicate with any other steps)?

This could work!

> or possibly it was required that the steps within a Task that is run hermetically inside a pod not communicate with each other? (probably a less good option since why would you have multiple steps at all unless they are communicating with each other - however we recently added [per step hermetic execution](https://github.com/tektoncd/community/blob/main/teps/0025-hermekton.md#api-changes) to the hermekton TEP which makes me think there is some value in having multiple steps even when they can't communicate with each other?)

It does seem to me like this would be less useful; what are the use cases motivating the per step hermetic execution? This option might be a bit confusing for users.

> Can multiple Tasks run in parallel in "pipeline in a pod" mode?
This is likely achievable but TBD if we want to do this. Requiring sequential execution if you want different levels of hermeticity (or just requiring sequential execution) is another reasonable option.

I'll add these options to the proposal.

Reply:
> This is likely achievable but TBD if we want to do this. Requiring sequential execution if you want different levels of hermeticity (or just requiring sequential execution) is another reasonable option.

I think the hermetic stuff **could work** here if there is no parallelism. If you allow parallel execution I don't think we can make any guarantees.

Reply:
Maybe we don't declare Hermekton in or out of scope for now, but leave this section documenting the constraints that need to be met on a group of tasks if we want to support hermetic execution? I don't have a good sense of whether users would value hermetic tasks or parallel execution (or multiple steps) more, or if they'd be ok with support for both but not at the same time.

Reply:
sounds good to me @lbernick !

>  Pipeline-in-a-pod would thus allow things to run in parallel.

(segue from the original topic)

There are some pipeline features which we already know we can't support in pipeline-in-a-pod (for example using a service account per task), I think we have some opportunity to decide which features we support in pipeline-in-a-pod and which we don't - especially in initial versions.

i'd suggest we leave parallel execution out of scope initially and see how useful pipeline-in-a-pod is without it, then determine how important it is to support it

Reply:
👍 I've removed Hermekton from non goals but kept in the design considerations.

Reply:
sgtm, thanks @lbernick !